cocotb/run.py

```diff
@@ -0,0 +1,184 @@
+    )
+
+    _build_lib(libgpilog, dist)
+
+    libcocotb = Extension(
+        "libcocotb",
+        define_macros=[("PYTHON_SO_LIB", python_lib)],
+        include_dirs=[share_dir + "/include"],
+        sources=[share_dir + "/lib/embed/gpi_embed.c"],
+    )
+
+    _build_lib(libcocotb, dist)
+
+    libgpi = Extension(
+        "libgpi",
+        include_dirs=[share_dir + "/include"],
+        libraries=["cocotbutils", "gpilog", "cocotb", "stdc++"],
+        library_dirs=[lib_path],
+        sources=[
+            share_dir + "/lib/gpi/GpiCbHdl.cpp",
+            share_dir + "/lib/gpi/GpiCommon.cpp",
+        ],
+    )
+
+    _build_lib(libgpi, dist)
+
+    libsim = Extension(
+        "simulator",
+        include_dirs=[share_dir + "/include"],
+        sources=[share_dir + "/lib/simulator/simulatormodule.c"],
+    )
+
+    _build_lib(libsim, dist)
+
+    libvpi = Extension(
+        "libvpi",
+        include_dirs=[share_dir + "/include"],
+        libraries=["gpi", "gpilog"],
+        library_dirs=[lib_path],
+        sources=[
+            share_dir + "/lib/vpi/VpiImpl.cpp",
+            share_dir + "/lib/vpi/VpiCbHdl.cpp",
+        ],
+    )
+
+    _build_lib(libvpi, dist)
+
+    _symlink_force(
+        os.path.join(lib_path, "libvpi." + ext_name),
+        os.path.join(lib_path, "gpivpi.vpl"),
+    )
+    _symlink_force(
+        os.path.join(lib_path, "libvpi." + ext_name),
+        os.path.join(lib_path, "cocotb.vpl"),
+    )
+
+    return lib_path
+
+
+def Run(sources, toplevel, module=None):
+
+    libs_dir = build_libs()
+
+    previous_frame = inspect.currentframe().f_back
+    (run_module_filename, _, _, _, _) = inspect.getframeinfo(previous_frame)
+
+    run_dir_name = os.path.dirname(run_module_filename)
+    run_module_name = os.path.splitext(os.path.split(run_module_filename)[-1])[0]
+
+    if module is None:
+        module = run_module_name
+
+    my_env = os.environ
+    my_env["LD_LIBRARY_PATH"] = libs_dir + ":" + sysconfig.get_config_var("LIBDIR")
+
+    python_path = ":".join(sys.path)
+    my_env["PYTHONPATH"] = python_path + ":" + libs_dir
+    my_env["TOPLEVEL"] = toplevel
+    my_env["TOPLEVEL_LANG"] = "verilog"
+    my_env["COCOTB_SIM"] = "1"
+    my_env["MODULE"] = module
+
+    sim_build_dir = os.path.join(run_dir_name, "sim_build")
+    os.makedirs(sim_build_dir, exist_ok=True)
+    sim_comopile_file = os.path.join(sim_build_dir, "sim.vvp")
+
+    sources_abs = []
+    for src in sources:
+        sources_abs.append(os.path.normpath(os.path.join(run_dir_name, src)))
+
+    comp_cmd = [
+        "iverilog",
+        "-o",
+        sim_comopile_file,
+        "-D",
+        "COCOTB_SIM=1",
+        "-s",
+        toplevel,
+        "-g2012",
+    ] + sources_abs
+    print(comp_cmd)
+    print(" ".join(comp_cmd))
+    process = subprocess.check_call(comp_cmd)
+
+    cmd = ["vvp", "-M", libs_dir, "-m", "gpivpi", sim_comopile_file]
+    print(" ".join(cmd))
+    process = subprocess.check_call(cmd)
```

Comment:
That's the tricky part: having a way to do something similar for all simulators. You could have a look at edalize, a python library (used in fusesoc) to solve exactly this problem.

Reply:
Right, will check. Makefiles are not better ;-)

Reply:
If I am not mistaken edalize generates a Makefile and then runs make.  Something I would like to avoid if possible?. 

Something else would be to add cocotb integration to edalize so you can use the same definition for synthesis etc. (I guess only useful for FPGAs).

Reply:
The approach edalize takes depends a bit on the tool: sometimes it uses Makefiles as intermediate way to call the tool, sometimes it doesn't. Actually the current thinking is to get rid of the Makefiles again as they make things harder on Windows. But at the same time, Windows development moves in a direction that makes the use of Unix tools easier, so there are probably good reasons to go either way.


Beyond that, integrating cocotb with fusesoc has been on my todo list for ages, and that's definitely something I'll pick up again soon.